docs/chymyst07.md

```diff
@@ -728,4 +728,104 @@ In the chemical machine, input molecules to each reaction must be defined static
 Since `n` is a run-time parameter, we cannot generalize to `n` in this way.
 
 
+## Cigarette smokers
+
+The "Cigarette smokers" problem is to implement four concurrent processes that coordinate assembly line operations to manufacture cigarettes with the 
+workers smoking the individual cigarettes they manufacture. One process represents a supplier of ingredients on the assembly line, which we may call the 
+pusher. The other processes are three smokers, each having an infinite supply of only one of the three required ingredients, which are matches, paper, and 
+tobacco. We may give names to the smokers/workers as Keith, Slash, and Jimi. 
+
+The pusher provides at random time intervals one unit each of two required ingredients (for example matches and paper, or paper and tobacco). The pusher is not allowed to coordinate with the smokers to use knowledge of which smoker 
+needs which ingredients, he just supplies two ingredients at a time on a conveyor belt. We assume that the pusher has an infinite stream of the three 
+ingredients available to him. The real life example is that of a core operating systems service having to schedule and provide limited distinct resources to 
+other services where coordination of scarce resources is required.
+ 
+Each smoker selects two ingredients, rolls up a cigarette using the third ingredient that complements the list, lits it up and smokes it. It is necessary
+ for the smoker to finish his cigarette before the pusher supplies the next two ingredients (a timer can simulate the smoking activity).
+
+
+We model the processes as follows:
+
+| Supplier   | | Smoker 1   | |  Smoker 2    | |  Smoker 3 |
+| --- | --- | --- | --- | --- | --- |  --- |
+| select 2 random ingredients | | pick tobacco and paper | | pick tobacco and matches | | pick matches and paper |
+
+
+Let us now figure out the chemistry that will solve this problem. We can think of the problem as a concurrent producer consumer queue with three competing 
+consumers and the supplier simply produces random pairs of ingredients into the queue. For simplicity, we assume the queue has capacity 1, which is an 
+assumption in the statement of the problem.
+
+It also helps to draw parallels with the Dining Philosophers problem realizing that we can think of three philosophers as consumers and the constraint on 
+matching up the three ingredients differently for the three smokers is a bit analog to the 
```

Comment:
That line didn't belong, at some point I had drawn a parallel between ingredients and forks.